docs/usage/jssengine.md

```diff
@@ -0,0 +1,521 @@
+# `JSSEngine` - Documentation
+
+## About `JSSEngine`
+
+`JSSEngine` is JSS's implementation of the [`SSLEngine`][javax.ssl-engine]
+contract for non-blocking TLS. Unlike the `SunJSSE` provider's `SSLEngine`
+(when using the `SunPKCS11-NSS` provider for primitives), this is built
+directly on NSS's high-level SSL module. This is better for FIPS compliance
+and HSM support (as keys never leave the NSS cryptographic module) and also
+means that we don't need to reimplement the underlying state machine. This
+approach is consistent with our [JSS `SSLSocket`][jss.ssl-socket].
+
+
+## Using `JSSEngine`
+
```

Comment:
This is great so far, but below it talks about the various ways to create an SSLEngine and configure it. Would it make any sense, in your opinion to put a little more up top as to why one would create one of these things to do what? Maybe a skeleton example cllient / server program. Of course to be fair, if someone wants to use this thing, they may possibly know already... Right now it just briefly describes that it is implements the contract for non-blocking TLS. Just a thought.

Reply:
An interesting idea, yes. Outside of test cases, I'm not sure most people will use a SSLEngine directly for client/server handshakes. There's a number of higher-level APIs (Tomcat, Apache products, and some internal to the JDK) that take a SSLEngine instance and use it to do, say, TLS for HTTPS or LDAPS. In my opinion, most people probably won't be using SSLEngine directly to do client/server ops and just need to "configure" it to pass it to the relevant library/...

Should I just provide a link to the [test cases](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/tests/TestSSLEngine.java) where this is done? Otherwise, there is an [excellent Oracle example](https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/samples/sslengine/SSLEngineSimpleDemo.java) that explains this as well.

Reply:
Yeah, the links sound good, but maybe just put a blurb or two explaining what this thing is for and how mostly it will be associated with higher level API's, just like you said and that should do it. Otherwise I"m just about finishing reading it and it looks great so far.